src/main/host/descriptor/descriptor_table.rs

```diff
@@ -0,0 +1,204 @@
+use super::CompatDescriptor;
+use crate::cshadow;
+use crate::utility::notnull::*;
+use log::*;
+use std::convert::TryInto;
+
+/// Table of (file) descriptors. Typically owned by a Process.
+pub struct DescriptorTable {
+    descriptors: std::collections::HashMap<u32, CompatDescriptor>,
+    available_indices: Vec<u32>,
+    index_counter: u32,
+}
+
+impl DescriptorTable {
+    pub fn new() -> Self {
+        DescriptorTable {
+            descriptors: std::collections::HashMap::new(),
+            available_indices: Vec::new(),
+            index_counter: 0,
+        }
+    }
+
+    /// Add the descriptor at an unused index, and return the index.
+    pub fn add(&mut self, descriptor: CompatDescriptor) -> u32 {
+        let mut descriptor = descriptor;
+
+        let idx = loop {
+            let idx = match self.available_indices.pop() {
+                Some(i) => {
+                    trace!("Reusing index {}", i);
+                    i
+                }
+                None => {
+                    let i = self.index_counter;
+                    self.index_counter += 1;
+                    trace!("New index {}", i);
+                    i
+                }
+            };
+            if self.descriptors.contains_key(&idx) {
+                // Already in use. This can happen after having called `set`,
+                // since it doesn't update `index_counter` or
+                // `available_indices`.  Assuming this is rare, cheapest
+                // solution in memory and CPU is to just loop here.
+                continue;
+            }
+            break idx;
+        };
+
+        descriptor.set_handle(idx);
+        let prev = self.descriptors.insert(idx, descriptor);
+        debug_assert!(prev.is_none(), "Already a descriptor at {}", idx);
+
+        idx
+    }
+
+    /// Remove the descriptor at the given index and return it.  Panics if there
+    /// is no descriptor at `idx`.
+    pub fn remove(&mut self, idx: u32) -> CompatDescriptor {
+        let mut descriptor = self.descriptors.remove(&idx).unwrap();
+        self.available_indices.push(idx);
+        descriptor.set_handle(0);
+        descriptor
+    }
+
+    /// Get the descriptor at `idx`, if any.
+    pub fn get(&self, idx: u32) -> Option<&CompatDescriptor> {
+        self.descriptors.get(&idx)
+    }
+
+    /// Insert a descriptor at `index`. If a descriptor is already present at
+    /// that index, it is unregistered from that index and returned.
+    pub fn set(&mut self, index: u32, descriptor: CompatDescriptor) -> Option<CompatDescriptor> {
+        let mut descriptor = descriptor;
```

Comment:
I thought it'd be a little weird to have `mut` in the interface, since the caller shouldn't care. I checked `cargo doc` though, and it looks like the generated docs do ignore it, so it's mostly moot :). Done.